fuzy-norms/script.whizzml

```diff
@@ -0,0 +1,232 @@
+  (when (< p -1)
+    (raise {"message" "Parameter for sugeno-weber t-norm must be >= -1"
+            "code" -3}))
+  (cond (= p -1)
+        (lambda(fid1 fid2)
+          {"field" (get (drastic-tnorm fid1 fid2) "field")
+           "name" (str "sugeno-weber-tnorm-p" p "-" fid1 "-" fid2)})
+        (lambda (fid1 fid2)
+          {"field" (flatline "(let (x (f {{fid1}}) y (f {{fid2}}))  "
+                             "  (max 0                                  "
+                             "       (/ (+ x y -1 (* {{p}} x y))        "
+                             "          (+ 1 {{p}}))))")
+           "name" (str "sugeno-weber-tnorm-p" p "-" fid1 "-" fid2)})))
+
+;; T-Conorms
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (max-tconorm fid1 fid2)
+  {"field" (flatline "(max (f {{fid1}}) (f {{fid2}}))")
+   "name" (str "max-tconorm-" fid1 "-" fid2)})
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (probabilistic fid1 fid2)
+  {"field" (flatline "(- (+ (f {{fid1}}) (f {{fid2}})  )"
+                     "   (* (f {{fid1}}) (f {{fid2}}))) ")
+   "name" (str "probabilistic-tconorm-" fid1 "-" fid2)})
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (bounded fid1 fid2)
+  {"field" (flatline "(min (+ (f {{fid1}}) (f {{fid2}})) 1)")
+   "name" (str "bounded-tconorm-" fid1 "-" fid2)})
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (drastic-tconorm fid1 fid2)
+  {"field" (flatline "(cond (= (f {{fid1}}) 0) (f {{fid2}}) "
+                     "      (= (f {{fid2}}) 0) (f {{fid1}})"
+                     "1.0)")
+   "name" (str "drastic-tconorm-" fid1 "-" fid2)})
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (nilpotent-max fid1 fid2)
+  {"field" (flatline "(if (< (+ (f {{fid1}}) (f {{fid2}})) 1)"
+                     " (max (f {{fid1}}) (f {{fid2}}))"
+                     " 1.0)")
+   "name" (str "nilpotent-max-tconorm-" fid1 "-" fid2)})
+
+;; https://en.wikipedia.org/wiki/T-norm
+(define (einstein-sum fid1 fid2)
+  {"field" (flatline "(/ (+ (f {{fid1}}) (f {{fid2}}))        "
+                     "   (+ 1 (* (f {{fid1}}) (f {{fid2}})))) ")
+   "name" (str "einstein-tconorm-" fid1 "-" fid2)})
+
+(define function-names
+  {"min-tnorm" min-tnorm
+   "product" product
+   "lukasiewicz" lukasiewicz
+   "drastic-tnorm" drastic-tnorm
+   "nilpotent-min" nilpotent-min
+   "schweizer-sklar" schweizer-sklar
+   "hamacher" hamacher
+   "yager" yager
+   "aczel-alsina" aczel-alsina
+   "dombi" dombi
+   "suge-weber" sugeno-weber
+   "max-tconorm" max-tconorm
+   "probabilistic" probabilistic
+   "bounded" bounded
+   "drastic-tconorm" drastic-tconorm
+   "nilpotent-max" nilpotent-max
+   "einstein-sum" einstein-sum})
+
+(define (not-found-error name)
+  (lambda (. opt)
+    (raise {"message" (str "Can't find norm " name)
+            "code" -3})))
+
+(define (check-function name)
+  (if (list? name)
+    ((get function-names (name 0) (not-found-error (name 0))) (name 1))
+    (get function-names name (not-found-error name))))
```

Comment:
you don't need `get`, `(function-names k d)` works and is more
idiomatic.  also, i would call this procedure `look-up-function`
or `look-up-norm` (it does not simply check, and it doesn't check
a function, but its name).